sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -91,6 +91,7 @@ case class AdaptiveSparkPlanExec(
   // optimizations should be stage-independent.
   @transient private val queryStageOptimizerRules: Seq[Rule[SparkPlan]] = Seq(
     ReuseAdaptiveSubquery(conf, subqueryCache),
+    OptimizedLocalShuffleReader(conf),
```

Comment:
since this may change the number of exchanges, we should put it in https://github.com/apache/spark/pull/25295/files#diff-6954dd8020a9ca298f1fb9602c0e831cR77

Then the AQE framework can check the cost and give up the optimization if extra changes are introduced.

Note that, making it a physical rule and check number of exchanges is suboptimal. It's possible that the local shuffle reader can avoid exchanges downstream, which changes the stage boundaries.

Reply:
Already move it in `queryStagePreparationRules`.